ocaml/testsuite/tests/typing-layouts-products/basics.ml

```diff
@@ -778,27 +778,15 @@ type t3 = #(int * bool) array
 type t4 = #(string * #(float# * bool option)) array
 |}]
 
-(* CR layouts v7.1: This example demonstrates a bug in type inference that we
-   should fix.  The issue is the way typing of tuple expressions works - we
-   create type variables at generic level and then constrain them by the
-   expected type.  This will fail if it requires to refine the kind, because we
-   don't allow refining kinds at generic level.  We need to remove the
-   restriction that kinds of things at generic level can't be modified, but that
-   is orthogonal to unboxed tuples so we leave in this sad bug for now. *)
+(* CR layouts v7.1: This should be accepted, or the error message should
+   be improved. *)
 let _ = [| #(1,2) |]
 [%%expect{|
-Line 1, characters 11-17:
+Line 1, characters 8-20:
 1 | let _ = [| #(1,2) |]
-               ^^^^^^
-Error: This expression has type #('a * 'b)
-       but an expression was expected of type
-         ('c : '_representable_layout_397 & '_representable_layout_398)
-       The kind of #('a * 'b) is
-         '_representable_layout_397 & '_representable_layout_398
-         because it is an unboxed tuple.
-       But the kind of #('a * 'b) must be a subkind of
-         '_representable_layout_397 & '_representable_layout_398
-         because it's the type of an array element.
+            ^^^^^^^^^^^^
+Error: Product layout value & value detected in structure in [Typeopt.Layout]
+       Please report this error to the Jane Street compilers team.
```

Comment:
How should the CR be changed? There's already a CR layouts v7.1. Given that you're the one most likely to pull that level, what comment would make you find it?

For what it's worth, I think it's safe to let this remain. Not a great UI, but not dangerous.

Reply:
I think if we roll to stable with this error, we'll get a lot of people trying unboxed tuples in arrays and reporting this, which would be good to avoid.  But I agree the CR is fine for now.